File: apps/nongyu-go-track-server/internal/aggregate/jobs.go
package aggregate

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"log/slog"
	"sort"
	"time"

	"nongyu-go-track-server/internal/bizday"
	"nongyu-go-track-server/internal/store/sqlite"
)

type Jobs struct {
	store *sqlite.Store
	log   *slog.Logger
	now   func() time.Time
	stop  chan struct{}
	done  chan struct{}

	lastAggMin   int
	lastPurgeMin int
}

func New(store *sqlite.Store, log *slog.Logger) *Jobs {
	return &Jobs{
		store:        store,
		log:          log,
		now:          time.Now,
		stop:         make(chan struct{}),
		done:         make(chan struct{}),
		lastAggMin:   -1,
		lastPurgeMin: -1,
	}
}

func (j *Jobs) Start() {
	go j.loop()
}

func (j *Jobs) Stop() {
	close(j.stop)
	<-j.done
}

func (j *Jobs) loop() {
	defer close(j.done)
	j.startupBackfill()
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-j.stop:
			return
		case <-ticker.C:
			j.tick()
		}
	}
}

func (j *Jobs) startupBackfill() {
	yesterday := bizday.Yesterday(j.now())
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	st, ok, err := j.store.JobStatus(ctx, "aggregate_daily", "stat_date="+yesterday)
	if err != nil {
		j.log.Error("check aggregate job", "err", err)
		return
	}
	if ok && st == "success" {
		return
	}
	if _, err := j.RunAggregate(ctx, yesterday); err != nil {
		j.log.Error("startup aggregate failed", "date", yesterday, "err", err)
	}
}

func (j *Jobs) tick() {
	now := j.now().In(bizday.Location())
	minute := now.Hour()*60 + now.Minute()
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	if now.Hour() == 0 && now.Minute() == 10 && j.lastAggMin != minute {
		j.lastAggMin = minute
		date := bizday.Yesterday(now)
		if _, err := j.RunAggregate(ctx, date); err != nil {
			j.log.Error("scheduled aggregate failed", "date", date, "err", err)
		}
	}
	if now.Hour() == 3 && now.Minute() == 0 && j.lastPurgeMin != minute {
		j.lastPurgeMin = minute
		if _, err := j.RunPurge(ctx); err != nil {
			j.log.Error("scheduled purge failed", "err", err)
		}
	}

	n, err := j.store.CountOnline(ctx)
	if err != nil {
		j.log.Warn("online sample failed", "err", err)
		return
	}
	if err := j.store.BumpPeak(ctx, bizday.StatDate(now), n, now.UTC().UnixMilli()); err != nil {
		j.log.Warn("online peak bump failed", "err", err)
	}
}

func (j *Jobs) RunAggregate(ctx context.Context, date string) (string, error) {
	key := "stat_date=" + date
	nowMs := j.now().UTC().UnixMilli()
	err := j.store.WithWriteTx(ctx, func(tx *sql.Tx) error {
		dau, err := j.store.CountDistinctDAU(ctx, tx, date)
		if err != nil {
			return err
		}
		appOpen, err := j.store.CountByType(ctx, tx, date, "app_open")
		if err != nil {
			return err
		}
		screens, err := j.store.CountByType(ctx, tx, date, "screen_view")
		if err != nil {
			return err
		}
		clicks, err := j.store.CountByType(ctx, tx, date, "button_click")
		if err != nil {
			return err
		}
		crashes, err := j.store.CountByType(ctx, tx, date, "crash")
		if err != nil {
			return err
		}
		if err := j.store.UpsertMetric(ctx, tx, date, "dau", dau, nowMs); err != nil {
			return err
		}
		if err := j.store.UpsertMetric(ctx, tx, date, "app_open_count", appOpen, nowMs); err != nil {
			return err
		}
		if err := j.store.UpsertMetric(ctx, tx, date, "screen_view_count", screens, nowMs); err != nil {
			return err
		}
		if err := j.store.UpsertMetric(ctx, tx, date, "button_click_count", clicks, nowMs); err != nil {
			return err
		}
		if err := j.store.UpsertMetric(ctx, tx, date, "crash_count", crashes, nowMs); err != nil {
			return err
		}

		screenDims, err := j.store.CountByName(ctx, tx, date, "screen_view")
		if err != nil {
			return err
		}
		if err := j.store.ReplaceDims(ctx, tx, date, "screen_views", nowMs, screenDims); err != nil {
			return err
		}
		clickDims, err := j.store.CountByName(ctx, tx, date, "button_click")
		if err != nil {
			return err
		}
		if err := j.store.ReplaceDims(ctx, tx, date, "button_clicks", nowMs, clickDims); err != nil {
			return err
		}

		perf, err := j.store.PerfDurations(ctx, tx, date)
		if err != nil {
			return err
		}
		p50 := make([]sqlite.DimRow, 0)
		p95 := make([]sqlite.DimRow, 0)
		for name, vals := range perf {
			sort.Slice(vals, func(i, k int) bool { return vals[i] < vals[k] })
			p50 = append(p50, sqlite.DimRow{DimKey: "name", DimValue: name, MetricValue: sqlite.Percentile(vals, 50)})
			p95 = append(p95, sqlite.DimRow{DimKey: "name", DimValue: name, MetricValue: sqlite.Percentile(vals, 95)})
		}
		if err := j.store.ReplaceDims(ctx, tx, date, "perf_p50", nowMs, p50); err != nil {
			return err
		}
		return j.store.ReplaceDims(ctx, tx, date, "perf_p95", nowMs, p95)
	})
	status := "success"
	detail := ""
	if err != nil {
		status = "failed"
		b, _ := json.Marshal(map[string]string{"error": err.Error()})
		detail = string(b)
		_ = j.store.RecordJob(ctx, "aggregate_daily", key, status, nowMs, detail)
		return status, err
	}
	if recErr := j.store.RecordJob(ctx, "aggregate_daily", key, status, nowMs, detail); recErr != nil {
		return status, recErr
	}
	return status, nil
}

func (j *Jobs) RunPurge(ctx context.Context) (int64, error) {
	cutoff := j.now().UTC().Add(-30 * 24 * time.Hour).UnixMilli()
	n, err := j.store.PurgeEvents(ctx, cutoff)
	nowMs := j.now().UTC().UnixMilli()
	key := fmt.Sprintf("at=%d", nowMs)
	status := "success"
	detail := ""
	if err != nil {
		status = "failed"
		b, _ := json.Marshal(map[string]string{"error": err.Error()})
		detail = string(b)
		_ = j.store.RecordJob(ctx, "purge_events", key, status, nowMs, detail)
		return 0, err
	}
	b, _ := json.Marshal(map[string]int64{"deleted": n})
	detail = string(b)
	if recErr := j.store.RecordJob(ctx, "purge_events", key, status, nowMs, detail); recErr != nil {
		return n, recErr
	}
	return n, nil
}

File: apps/nongyu-node-track-server/src/aggregate/jobs.ts
import type { Logger } from "pino";
import { shanghaiHourMinute, statDate, yesterday } from "../bizday.js";
import type { Store } from "../store/sqlite/db.js";
import {
  avgScreenDwell,
  bumpPeak,
  countByName,
  countByType,
  countDistinctDAU,
  countScreenEnters,
  jobStatus,
  percentile,
  perfDurations,
  purgeEvents,
  recordJob,
  replaceDims,
  upsertMetric,
} from "../store/sqlite/metrics.js";
import { countOnline } from "../store/sqlite/presence.js";

export class Jobs {
  private stopped = false;
  private loopPromise: Promise<void> | null = null;
  private lastAggMin = -1;
  private lastPurgeMin = -1;
  now: () => Date = () => new Date();

  constructor(
    private store: Store,
    private log: Logger,
  ) {}

  start(): void {
    this.loopPromise = this.loop();
  }

  async stop(): Promise<void> {
    this.stopped = true;
    if (this.loopPromise) await this.loopPromise;
  }

  private async loop(): Promise<void> {
    this.startupBackfill();
    while (!this.stopped) {
      await sleep(30_000);
      if (this.stopped) break;
      this.tick();
    }
  }

  private startupBackfill(): void {
    const date = yesterday(this.now());
    try {
      const st = jobStatus(this.store, "aggregate_daily", `stat_date=${date}`);
      if (st.ok && st.status === "success") return;
      void this.runAggregate(date);
    } catch (err) {
      this.log.error({ err, date }, "startup aggregate failed");
    }
  }

  private tick(): void {
    const now = this.now();
    const { hour, minute } = shanghaiHourMinute(now);
    const minuteOfDay = hour * 60 + minute;

    if (hour === 0 && minute === 10 && this.lastAggMin !== minuteOfDay) {
      this.lastAggMin = minuteOfDay;
      const date = yesterday(now);
      try {
        this.runAggregate(date);
      } catch (err) {
        this.log.error({ err, date }, "scheduled aggregate failed");
      }
    }
    if (hour === 3 && minute === 0 && this.lastPurgeMin !== minuteOfDay) {
      this.lastPurgeMin = minuteOfDay;
      try {
        this.runPurge();
      } catch (err) {
        this.log.error({ err }, "scheduled purge failed");
      }
    }

    try {
      const n = countOnline(this.store);
      bumpPeak(this.store, statDate(now), n, now.getTime());
    } catch (err) {
      this.log.warn({ err }, "online peak bump failed");
    }
  }

  runAggregate(date: string): { status: string } {
    const key = `stat_date=${date}`;
    const nowMs = this.now().getTime();
    try {
      this.store.withWriteTx(() => {
        const appOnly = { excludePlatform: "web" as const };
        const dau = countDistinctDAU(this.store, date);
        const appOpen = countByType(this.store, date, "app_open");
        const screensAll = countByType(this.store, date, "screen_view");
        const webScreens = countByType(this.store, date, "screen_view", "web");
        const screens = screensAll - webScreens;
        const clicks = countByType(this.store, date, "button_click");
        const crashes = countByType(this.store, date, "crash");
        upsertMetric(this.store, date, "dau", dau, nowMs);
        upsertMetric(this.store, date, "app_open_count", appOpen, nowMs);
        upsertMetric(this.store, date, "screen_view_count", screens, nowMs);
        upsertMetric(this.store, date, "button_click_count", clicks, nowMs);
        upsertMetric(this.store, date, "crash_count", crashes, nowMs);

        replaceDims(
          this.store,
          date,
          "screen_views",
          nowMs,
          countScreenEnters(this.store, date, appOnly),
        );
        replaceDims(
          this.store,
          date,
          "screen_dwell_avg",
          nowMs,
          avgScreenDwell(this.store, date, appOnly),
        );
        replaceDims(
          this.store,
          date,
          "button_clicks",
          nowMs,
          countByName(this.store, date, "button_click", appOnly),
        );

        const perf = perfDurations(this.store, date, appOnly);
        const p50: Array<{ dimKey: string; dimValue: string; metricValue: number }> = [];
        const p95: Array<{ dimKey: string; dimValue: string; metricValue: number }> = [];
        for (const [name, vals] of perf) {
          const sorted = [...vals].sort((a, b) => a - b);
          p50.push({ dimKey: "name", dimValue: name, metricValue: percentile(sorted, 50) });
          p95.push({ dimKey: "name", dimValue: name, metricValue: percentile(sorted, 95) });
        }
        replaceDims(this.store, date, "perf_p50", nowMs, p50);
        replaceDims(this.store, date, "perf_p95", nowMs, p95);
      });
      recordJob(this.store, "aggregate_daily", key, "success", nowMs, "");
      return { status: "success" };
    } catch (err) {
      const detail = JSON.stringify({ error: err instanceof Error ? err.message : String(err) });
      try {
        recordJob(this.store, "aggregate_daily", key, "failed", nowMs, detail);
      } catch {
        /* ignore */
      }
      throw err;
    }
  }

  runPurge(): { deleted: number } {
    const cutoff = this.now().getTime() - 30 * 24 * 60 * 60 * 1000;
    const nowMs = this.now().getTime();
    const key = `at=${nowMs}`;
    try {
      const n = purgeEvents(this.store, cutoff);
      recordJob(this.store, "purge_events", key, "success", nowMs, JSON.stringify({ deleted: n }));
      return { deleted: n };
    } catch (err) {
      const detail = JSON.stringify({ error: err instanceof Error ? err.message : String(err) });
      try {
        recordJob(this.store, "purge_events", key, "failed", nowMs, detail);
      } catch {
        /* ignore */
      }
      throw err;
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
